File: 0227-basic-calculator-ii/0227-basic-calculator-ii.java
class Solution {
    public int calculate(String s) {
        s = s.trim();
        int curr = 0; // main curr number
        int calc = 0;
        Stack<Integer> st = new Stack<>();
        char lastSign = '+';

        for(int i = 0; i < s.length(); i++){
            char ch = s.charAt(i);

            if(Character.isDigit(ch)){
                curr = curr * 10 + ch - '0';
            }

            if(ch != ' ' && (!Character.isDigit(ch) || i == s.length() - 1)){
                if(lastSign == '+'){
                    st.push(curr);
                }
                else if(lastSign == '-'){
                    st.push(-curr);
                }
                else if(lastSign == '*'){
                    st.push(st.pop() * curr);
                }
                else if(lastSign == '/'){
                    st.push(st.pop() / curr);
                }

                lastSign = ch;
                curr = 0;
            }
        }

        while(!st.empty()){
            calc += st.pop();
        }
        return calc;
    }
}
// Stack
// Time Complexity: O(n)
// Space Complexity: O(n)

File: 0227-basic-calculator-ii/0227-basic-calculator-ii.py
class Solution:
    def calculate(self, s: str) -> int:
        if s is None or len(s) == 0:
            return 0
        
        # tail = 0
        curr = 0
        calc = 0
        # s.strip()

        lastSign = '+'
        st = []

        for i in range(len(s)):
            ch = s[i]

            if ch.isnumeric():
                curr = curr * 10 + int(ch)
            
            if(not ch.isdigit() and ch != " ") or i == len(s) - 1:
                if lastSign == "+":
                    st.append(curr)
                elif lastSign == "-":
                    st.append(-curr)
                elif lastSign == "*":                    
                    st.append(st.pop() * curr)
                elif lastSign == "/":
                    st.append(int(st.pop() / curr))
                
                curr = 0
                lastSign = ch
        
        while len(st) > 0:
            calc += st.pop()
        return calc

# Time Complexity: O(n)
# Space Complexity: O(n)